st2client/st2client/utils/jsutil.py

```diff
@@ -13,19 +13,49 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from jsonpath_rw import parse
-
 
 def get_value(doc, key):
```

Comment:
Can you please add some tests for this function? Or we have some already?

Reply:
@Kami they already exist here : https://github.com/StackStorm/st2/blob/master/st2client/tests/unit/test_util_json.py

Reply:
As per https://github.com/StackStorm/st2/pull/3854#pullrequestreview-75952944 it looks like that a good start should be, in any case, to add tests for all the various edge case which it looks like we don't have right now (dictionary key contains a dot, etc.) - it looks like we have just a couple of very basic / simple test cases.

This way we have a good starting ground and no matter the solution we go with (aka the implementation detail which uses jsonpath-rw right now) there will be a smaller chance for regression, at least for the scenarios we have tests for.

So in short, I would start with test cases for various common and edge cases for `get_value` and work there.

If we can be reasonably sure that there are no regressions and that all the tests still pass with your proposed implementation, I would be fine with merging that.

/cc @Mierdin what do you think? I think jsonpath-rw is very flexible, but if we have good test coverage for all the common and edge cases we want to support, I don't care about the actual implementation detail.

Perhaps one option would also be to use this simple / naive approach when only "foo.bar.baz" notation is used and for any other more advanced queries, simple fall back to jsonpath-rw. I would be fine with that as a good and pragmatic compromise for a common case.

Reply:
And another future improvement / performance optimization would be to skip loading, parsing and dumping round-trip all together when user uses --json argument - https://github.com/StackStorm/st2/blob/master/st2client/st2client/formatters/doc.py#L34.

API already returns JSON and if user simply requests JSON output without any other flags such as `--k`, there is no need to go through the - parse, process, dump round-trip, it just provides no value and adds a lot of overhead.

Reply:
I believe the reason that is not done currently is that all of the fields are sent from the API to the Client and then the client does the filtering.

By default, for every "command" / type there are certain fields that are shown called `display_attributes`: https://github.com/StackStorm/st2/blob/master/st2client/st2client/commands/action.py#L200

Maybe a better improvement would be to perform these filters on the server side using the MongoDB API? I'm not very familiar with Mongo but my SQL statement would look something like: `SELECT foo, bar, baz FROM model_table`. This would decrease the serialization requirements and the size of the payload sent between the server and client.

Reply:
Back to the testing comment, I would love to see these tests do a little bit more. If we do end up with the "simple way" for common cases, with a fallback to jsonpath_rw, a "complicated case" example I would recommend is when IP addresses (with dotted decimal notation) are used as keys. Things like that, that would require more than a simple key.key.key lookup

Reply:
@Mierdin working on this today (finally... sorry it took so long)

But it looks like the current implementation doesn't work for IP address queries. The `get_kvps` function sees the `.` in the key and splits it up into nested dictionaries. I'm going to spend a little time trying to fix it, but it basically boils down to that `get_kvps` expects "simple" expressions also where's the `.` in keys denote separators for nested dictionaries.
Example:
```

DOC_IP_ADDRESS = {
    'ips': {
        "192.168.1.1": {
            "hostname": "router.domain.tld"
        },
        "192.168.1.10": {
            "hostname": "server.domain.tld"
        }
    }
}

print jsutil.get_kvps(DOC_IP_ADDRESS,
                                         ['ips."192.168.1.1"', 'ips."192.168.1.10".hostname'])

# {'ips': {'"192': {'168': {'1': {'1"': {'hostname': 'router.domain.tld'},
#                               '10"': {'hostname': 'server.domain.tld'}}}}}}
```